[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158692


**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a user-facing API `create_cp_block_mask` for creating CP-specific `BlockMask`
which masks over the attention result of Q shard and KV global.
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations.
- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)


What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

**Followup**
1. create an issue to reproduce the error in `create_fw_bw_graph()` when trying to call `create_block_mask`
to re-write `block_mask` in `FlexAttentionHOP` dispatch in `TorchFunctionMode`.
2. Merge `_ContextParallelGlobalVars` and `_cp_options`.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

XilunWu committed:
[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations, and `cp_block_mask` is the `BlockMask` to be applied to
`Attention(q_shard, k_global, v_global)`. This block mask cannot be created within
`TorchFunctionMode` otherwise leading to an error in `create_fw_bw_graph()`.
- add a util function `rewrite_context_parallel_block_mask` which given a `BlockMask`
over `q_global`, `k_global`, and `v_global` produces a `BlockMask` that applies to `q_shard`,
`k_global`, and `v_global`.
- 

What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

drisspg commented on test/distributed/tensor/test_attention.py:
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
+            mesh_shape=(self.world_size,),
+            mesh_dim_names=("cp",),
+        )
+        # NOTE: call create_block_mask() within TorchFunctionMode would cause error in create_fw_bw_graph
```
You created an issue for this right?

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1086,6 +1115,74 @@ def backward_hook(grad: torch.Tensor) -> None:
+            kv_idx,
+        )
+
+    def _create_block_mask(
```
Do we need this?

XilunWu replied:
good catch! I added this because I was planning to add `lru_cache` but actually this cache is not needed because I'm now requiring user to explicitly initialize the `BlockMask`s instead of implicitly `create_block_mask` in `FlexAttentionHOP` dispatch.

drisspg replied:
I typically create a global `create_block_mask_compiled` and use that one

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1139,6 +1283,36 @@ def __torch_function__(
 
+            # special handler for flex_attention
+            if func == torch._higher_order_ops.flex_attention:
+                # raise NotImplementedError("flex_attention is not supported yet.")
```
nit: remove the comments

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2793,6 +2793,11 @@ def call_function(
         args: "list[VariableTracker]",
         kwargs: "dict[str, VariableTracker]",
     ) -> "VariableTracker":
+        from .torch_function import can_dispatch_torch_function, dispatch_torch_function
+
+        if can_dispatch_torch_function(tx, args, kwargs):
+            return dispatch_torch_function(tx, self, args, kwargs)
+
```
seems reasonable to me, but (1) can we make sure that in eager-mode, flexattention has a torch_function and (2) can we test that in both eager and compile, flexattention torch_function gets dispatched?

XilunWu commented on test/distributed/tensor/test_attention.py:
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
+            # this is the block_mask created within the training step
+            # NOTE: flex_attention checks block_mask shape and input shape before
+            # calling into flex_attention_hop.
+            block_mask_post_sharding = create_block_mask(
```
sync with Driss offline: we can request users to explicitly call a new API `create_cp_block_mask()` that returns (3)rd `BlockMask` but the Q_LEN and K_LEN are modified to those in (2)nd, and reset to the right value right before passing to `FlexAttentionHOP.__call__` in `TorchFunctionMode`.

fegin replied:
Just want to confirm, is the proposal adding `create_cp_block_mask` to the CP APIs?

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations, and `cp_block_mask` is the `BlockMask` to be applied to
`Attention(q_shard, k_global, v_global)`. This block mask cannot be created within
`TorchFunctionMode` otherwise leading to an error in `create_fw_bw_graph()`.
- add a util function `rewrite_context_parallel_block_mask` which given a `BlockMask`
over `q_global`, `k_global`, and `v_global` produces a `BlockMask` that applies to `q_shard`,
`k_global`, and `v_global`.
- an Autograd Function `CPFlexAttentionPreOp` that performs all-gather over KV in forward and
reduce-scatter over grad_K and grad_V in backward. This is introduced because `TorchFunctionMode`
does not observe the backward HOP.

What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

fegin commented on test/distributed/tensor/test_attention.py:
```diff
@@ -437,5 +439,172 @@ def _test_ring_attention_custom_transformer(self, rotater: _RotateMethod) -> Non
+            self._test_ring_flex_attention,
+        )
+
+        # TODO: the current CP impl doesn't support small attentions (block_size < 128)
```
Is this CP's limitation? Where is the limitation from?

fegin commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -66,6 +71,29 @@ class _ContextParallelOptions:
 
 
+@dataclass
+class _ContextParallelGlobalVars:
```
Since we have this dataclass, should we move `_dispatch_mode` and `_cp_options` to this dataclass?

XilunWu replied:
Yes we can merge them. I'll do that in a follow-up PR (this requires torchtitan change as well).

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations, and `cp_block_mask` is the `BlockMask` to be applied to
`Attention(q_shard, k_global, v_global)`. This block mask cannot be created within
`TorchFunctionMode` otherwise leading to an error in `create_fw_bw_graph()`.
- add a util function `rewrite_context_parallel_block_mask` which given a `BlockMask`
over `q_global`, `k_global`, and `v_global` produces a `BlockMask` that applies to `q_shard`,
`k_global`, and `v_global`.
- an Autograd Function `CPFlexAttentionPreOp` that performs all-gather over KV in forward and
reduce-scatter over grad_K and grad_V in backward. This is introduced because `TorchFunctionMode`
does not observe the backward HOP.

What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations, and `cp_block_mask` is the `BlockMask` to be applied to
`Attention(q_shard, k_global, v_global)`. This block mask cannot be created within
`TorchFunctionMode` otherwise leading to an error in `create_fw_bw_graph()`.
- add a util function `rewrite_context_parallel_block_mask` which given a `BlockMask`
over `q_global`, `k_global`, and `v_global` produces a `BlockMask` that applies to `q_shard`,
`k_global`, and `v_global`.
- an Autograd Function `CPFlexAttentionPreOp` that performs all-gather over KV in forward and
reduce-scatter over grad_K and grad_V in backward. This is introduced because `TorchFunctionMode`
does not observe the backward HOP.

What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

yangw-dev commented:
@pytorchbot revert -m "failed some internal testapf.metrics.tests.generate_graph_def_test.GenerateGraphDefTest: test_aps_generate_inference_graph_def_with_justknobs1) AssertionError: Expected 'check' to be called once. Called 3 times., please fix the internal test and reland it"   -c ghfirst

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a user-facing API `create_cp_block_mask` for creating CP-specific `BlockMask`
which masks over the attention result of Q shard and KV global.
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations.
- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)


What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

**Followup**
1. create an issue to reproduce the error in `create_fw_bw_graph()` when trying to call `create_block_mask`
to re-write `block_mask` in `FlexAttentionHOP` dispatch in `TorchFunctionMode`.
2. Merge `_ContextParallelGlobalVars` and `_cp_options`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel][FlexAttention] Prototype of supporting FlexAttention in Context Parallel"



**Summary**
This PR adds an all-gather based FlexAttention and uses TorchFunctionMode to dispatch
`FlexAttentionHOP.__call__` to it.

This PR makes the following changes:

- add a user-facing API `create_cp_block_mask` for creating CP-specific `BlockMask`
which masks over the attention result of Q shard and KV global.
- add `_ContextParallelGlobalVars` to store all necessary global vars that CP FlexAttention
requires. `torch_function_mode` is critical to maintain singleton mode to avoid dynamo
recompilations.
- add a dispatch path for `FlexAttentionForwardHOP.__call__` (TorchFunctionMode dispatch
won't work correctly without this line)


What's not in this PR:
- QKV load balancing
- Test on other masking besides `causal_mask`.
- Support on small attention (i.e. qkv size is smaller than 128) because the block mask
rewrite function requires `Q_BLOCK_SIZE == KV_BLOCK_SIZE == 128`.

**Test**
`pytest test/distributed/tensor/test_attention.py -s -k test_ring_flex_attention`

**Followup**
1. create an issue to reproduce the error in `create_fw_bw_graph()` when trying to call `create_block_mask`
to re-write `block_mask` in `FlexAttentionHOP` dispatch in `TorchFunctionMode`.
2. Merge `_ContextParallelGlobalVars` and `_cp_options`.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
